File: packages/server/src/agent/market.ts
/** マルチマーケット対応の市場開閉判定 */

import { MARKETS, type MarketConfig } from "./markets.js";

function nowInTz(mkt: MarketConfig): Date {
  // Get current time in target timezone
  const now = new Date();
  const tzStr = now.toLocaleString("en-US", { timeZone: mkt.timezone });
  return new Date(tzStr);
}

export function isMarketOpen(marketId = "us"): boolean {
  const mkt = MARKETS[marketId];
  if (!mkt) return false;

  const now = nowInTz(mkt);

  // Weekend check
  if (now.getDay() === 0 || now.getDay() === 6) return false;

  // Holiday check
  const dateStr = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}-${String(now.getDate()).padStart(2, "0")}`;
  if (mkt.holidays.has(dateStr)) return false;

  // Market hours check
  const currentMinutes = now.getHours() * 60 + now.getMinutes();
  const openMinutes = mkt.openHour * 60 + mkt.openMinute;
  const closeMinutes = mkt.closeHour * 60 + mkt.closeMinute;

  if (currentMinutes < openMinutes || currentMinutes > closeMinutes) {
    return false;
  }

  // Lunch break check (TSE etc.)
  if (mkt.breakStartHour != null && mkt.breakStartMinute != null && mkt.breakEndHour != null && mkt.breakEndMinute != null) {
    const breakStart = mkt.breakStartHour * 60 + mkt.breakStartMinute;
    const breakEnd = mkt.breakEndHour * 60 + mkt.breakEndMinute;
    if (currentMinutes >= breakStart && currentMinutes <= breakEnd) {
      return false;
    }
  }

  return true;
}

export function isTradingDay(marketId = "us"): boolean {
  const mkt = MARKETS[marketId];
  if (!mkt) return false;

  const now = nowInTz(mkt);

  if (now.getDay() === 0 || now.getDay() === 6) return false;

  const dateStr = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}-${String(now.getDate()).padStart(2, "0")}`;
  return !mkt.holidays.has(dateStr);
}

File: agent/src/agent_runner/market.py
"""マルチマーケット対応の市場開閉判定"""

from datetime import datetime

import pytz

from agent_runner.markets import MARKETS, MarketConfig


def _now_in_tz(mkt: MarketConfig) -> datetime:
    return datetime.now(pytz.timezone(mkt.timezone))


def is_market_open(market_id: str = "us") -> bool:
    mkt = MARKETS[market_id]
    now = _now_in_tz(mkt)

    if now.weekday() >= 5:
        return False
    if now.strftime("%Y-%m-%d") in mkt.holidays:
        return False

    market_open = now.replace(hour=mkt.open_hour, minute=mkt.open_minute, second=0, microsecond=0)
    market_close = now.replace(hour=mkt.close_hour, minute=mkt.close_minute, second=0, microsecond=0)

    if not (market_open <= now <= market_close):
        return False

    # 昼休みチェック (TSE 等)
    if mkt.break_start_hour is not None:
        break_start = now.replace(hour=mkt.break_start_hour, minute=mkt.break_start_minute, second=0, microsecond=0)
        break_end = now.replace(hour=mkt.break_end_hour, minute=mkt.break_end_minute, second=0, microsecond=0)
        if break_start <= now <= break_end:
            return False

    return True


def is_trading_day(market_id: str = "us") -> bool:
    mkt = MARKETS[market_id]
    now = _now_in_tz(mkt)
    if now.weekday() >= 5:
        return False
    return now.strftime("%Y-%m-%d") not in mkt.holidays
